File: solutions/cpp/protein-translation/1/protein_translation.cpp
#include "protein_translation.h"

#include <vector>
#include <string>
#include <map>

using namespace std;

namespace protein_translation {

    map<string,string> amino_map = {
        {"AUG", "Methionine"},
        {"UUU", "Phenylalanine"},
        {"UUC", "Phenylalanine"},
        {"UUA", "Leucine"},
        {"UUG", "Leucine"},
        {"UCU", "Serine"},
        {"UCC", "Serine"},
        {"UCA", "Serine"},
        {"UCG", "Serine"},
        {"UAU", "Tyrosine"},
        {"UAC", "Tyrosine"},
        {"UGU", "Cysteine"},
        {"UGC", "Cysteine"},
        {"UGG", "Tryptophan"},
        {"UAA", "STOP"},
        {"UAG", "STOP"},
        {"UGA", "STOP"}    
    };
    

    vector<string> proteins(string rna) {
            vector<string> result;
            for(int i=0; i<=static_cast<int>(rna.length()-3); i+=3){
                string codon = rna.substr(i,3);
                auto search = amino_map.find(codon);
                if(search!=amino_map.end()){
                    if(search->second == "STOP")
                        break;
                    else
                        result.push_back(search->second);
                } 
            }
        return result;
    }

}  // namespace protein_translation

File: solutions/python/protein-translation/1/protein_translation.py
amino_map = {"AUG":"Methionine",  
               "UUU":"Phenylalanine",
               "UUC":"Phenylalanine",
               "UUA":"Leucine",
               "UUG":"Leucine",
               "UCU":"Serine",
               "UCC":"Serine",
               "UCA":"Serine",
               "UCG":"Serine",
               "UAU":"Tyrosine",
               "UAC":"Tyrosine",
               "UGU":"Cysteine",
               "UGC":"Cysteine",
               "UGG":"Tryptophan",
               "UAA":"STOP",
               "UAG":"STOP",
               "UGA":"STOP"}

def triplets(strand):
    return [strand[i:i+3] for i in range(0, len(strand),3)]

def proteins(strand):
    polypeptide = []
    trips = triplets(strand)
    for triplet in trips:
        if triplet in amino_map:
            amino = amino_map[triplet]
            if amino=="STOP":
                return polypeptide
            else:
                polypeptide.append(amino)
    return polypeptide
    
